src/main/java/org/jenkinsci/plugins/workflow/cps/CpsGroovyShell.java

```diff
@@ -28,9 +35,51 @@
     /**
      * Use {@link CpsGroovyShellFactory} to instantiate it.
      */
+    @SuppressFBWarnings(value="DP_CREATE_CLASSLOADER_INSIDE_DO_PRIVILEGED", justification="irrelevant")
     CpsGroovyShell(ClassLoader parent, @CheckForNull CpsFlowExecution execution, CompilerConfiguration cc) {
         super(parent,new Binding(),cc);
         this.execution = execution;
+        try {
+            // Apparently there is no supported way to initialize a GroovyShell with a specified GroovyClassLoader.
+            Field loaderF = GroovyShell.class.getDeclaredField("loader");
+            loaderF.setAccessible(true);
+            loaderF.set(this, new CleanGroovyClassLoader(parent, cc));
+        } catch (Exception x) {
+            LOGGER.log(Level.WARNING, "failed to install CleanGroovyClassLoader", x);
+        }
+    }
+
+    /**
+     * Disables the weird and unreliable {@link groovy.lang.GroovyClassLoader.InnerLoader}.
+     * This is apparently only necessary when you are using class recompilation, which we are not.
+     * We want the {@linkplain Class#getClassLoader defining loader} of {@code *.groovy} to be this one.
+     * Otherwise the defining loader will be an {@code InnerLoader}, and not necessarily the same instance from load to load.
+     * @see GroovyClassLoader#getTimeStamp
+     */
+    private static final class CleanGroovyClassLoader extends GroovyClassLoader {
```

Comment:
Is there something I'm missing here, or is this basically just stripping out the timestamp from InnerLoader?

There's a wrinkle here, which is that InnerLoader seems like it may support nesting, and I'm wondering if there are implications if that is removed.  But perhaps I misunderstand?